File: leetcode/python/0001-two-sum.py
# path: leetcode\python\0001-two-sum.py
# link https://leetcode.com/problems/two-sum/
# date 2023-04-07
# leetcode: Easy

from ast import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        prevMap = {} # val : index

        for i, n in enumerate(nums):
            diff = target - nums[i]
            if diff in prevMap:
                return [prevMap[diff], i]
            prevMap[n] = i

File: leetcode/typescript/0001-two-sum.ts
/**
 * Problem: 1. Two Sum
 * Time O(N) | Space O(N)
 * Path: leetcode\typescript\0001-two-sum.ts
 * Link: https://leetcode.com/problems/two-sum/
 * Leetcode: Easy
 * Date: 2023-05-08
 * Redoing all the easy problems from scratch to get used to TypeScript
 */

function twoSum(nums: number[], target: number): number[] {
  const map = new Map<number, number>();
  for (let i = 0; i < nums.length; i++) {
    const diff = target - nums[i];
    if (map.has(diff)) {
      return [map.get(diff), i];
    } else {
      map.set(nums[i], i);
    }
  }
  return [];
}
